src/hugr/views/sibling.rs

```diff
@@ -0,0 +1,562 @@
+        // TODO: handle state order edges. For now we ignore them entirely.
+        // See https://github.com/CQCL-DEV/hugr/discussions/432
+        let rep_inputs = replacement
+            .node_outputs(rep_input)
+            // filter out any non-dataflow ports
+            .filter(|&p| {
+                replacement
+                    .get_optype(rep_input)
+                    .signature()
+                    .get(p)
+                    .is_some()
+            });
+        let rep_outputs = replacement
+            .node_inputs(rep_output)
+            // filter out any non-dataflow ports
+            .filter(|&p| {
+                replacement
+                    .get_optype(rep_output)
+                    .signature()
+                    .get(p)
+                    .is_some()
+            });
+        let self_inputs = self.incoming_ports();
+        let self_outputs = self.outgoing_ports();
+        let nu_inp = rep_inputs
+            .zip_eq(self_inputs)
+            .flat_map(|(rep_source, self_target)| {
+                replacement
+                    .linked_ports(rep_input, rep_source)
+                    .map(move |rep_target| (rep_target, self_target))
+            })
+            .collect();
+        let nu_out = self_outputs
+            .zip_eq(rep_outputs)
+            .flat_map(|((self_source_n, self_source_p), rep_target)| {
+                self.base
+                    .linked_ports(self_source_n, self_source_p)
+                    .map(move |self_target| (self_target, rep_target))
+            })
+            .collect();
+
+        Ok(SimpleReplacement::new(
+            self.get_parent(),
+            removal,
+            replacement,
+            nu_inp,
+            nu_out,
+        ))
+    }
+}
+
+/// Errors that can occur while constructing a [`SimpleReplacement`].
+#[derive(Debug, Clone, PartialEq, Eq, Error)]
+pub enum InvalidReplacement {
+    /// No DataflowParent root in replacement graph.
+    #[error("No DataflowParent root in replacement graph.")]
+    InvalidDataflowGraph,
+    /// Malformed DataflowParent in replacement graph.
+    #[error("Malformed DataflowParent in replacement graph.")]
+    InvalidDataflowParent,
+    /// Replacement graph boundary size mismatch.
+    #[error("Replacement graph boundary size mismatch.")]
+    InvalidSignature,
+    /// SiblingSubgraph is not convex.
+    #[error("SiblingSubgraph is not convex.")]
+    NonConvexSubgraph,
+}
+
+/// Errors that can occur while constructing a [`SiblingSubgraph`].
+#[derive(Debug, Clone, PartialEq, Eq, Error)]
+pub enum InvalidSubgraph {
+    /// The subgraph is not convex.
+    #[error("The subgraph is not convex.")]
+    NotConvex,
+    /// Not all nodes have the same parent.
+    #[error("Not a sibling subgraph.")]
+    NoSharedParent,
+    /// Empty subgraphs are not supported.
+    #[error("Empty subgraphs are not supported.")]
+    EmptySubgraph,
+}
+
+// /// A common trait for views of a HUGR sibling subgraph.
```

Comment:
Probably just remove this? Unless you want to `impl` it for SiblingGraph and SiblingSubgraph as a no-methods marker trait